File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.java
class TrieNode{
    HashMap<Character, TrieNode> children = new HashMap<Character, TrieNode>();
    boolean endOfWord = false;

    public TrieNode(){

    }
}

class Trie {
    TrieNode root = new TrieNode();

    public Trie() {
        
    }
    
    public void insert(String word) {
        TrieNode curr = root;
        for (Character c: word.toCharArray()){
            if(!curr.children.containsKey(c)) curr.children.put(c, new TrieNode());
            curr = curr.children.get(c);
        }
        curr.endOfWord = true;
    }
    
    public boolean search(String word) {
        TrieNode curr = root;
        for (Character c: word.toCharArray()){
            if(!curr.children.containsKey(c)) return false;
            curr = curr.children.get(c);
        }
        return curr.endOfWord;
    }
    
    public boolean startsWith(String prefix) {
        TrieNode curr = root;
        for (Character c: prefix.toCharArray()){
            if(!curr.children.containsKey(c)) return false;
            curr = curr.children.get(c);
        }
        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.cpp
class TrieNode{
public:
    unordered_map<char, TrieNode*> children;
    bool endOfWord = false;
    TrieNode(){
    }
};

class Trie {
public:
    TrieNode* root = new TrieNode();

    Trie() {
    }
    
    void insert(string word) {
        TrieNode* curr = root;
        for (char c: word){
            if (!curr->children.contains(c)){
                curr->children[c] = new TrieNode();
            }
            curr = curr->children[c];
        }
        curr->endOfWord = true;
    }
    
    bool search(string word) {
        TrieNode* curr = root;
        for (char c: word){
            if (!curr->children.contains(c)) return false;
            curr = curr->children[c];
        }
        return curr->endOfWord;
    }
    
    bool startsWith(string prefix) {
        TrieNode* curr = root;
        for (char c: prefix){
            if (!curr->children.contains(c)) return false;
            curr = curr->children[c];
        }
        return true;
    }
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */